spec/core/v2/ics-004-channel-and-packet-semantics/README.md

```diff
@@ -0,0 +1,1028 @@
+| **Post-Conditions (Success)**  | 1. `onReceivePacket` is executed and the application state is modified<br> 2. The `packetReceipt` is written<br> 3. Event is Emitted<br>  | 1. `onReceivePacket(..)==True; app.State(beforeReceivePacket)!=app.State(afterReceivePacket)`<br> 2. `provableStore.get(packetReceiptPath(packet.channelDestId, packet.sequence))!=null`<br> 3. Check Event Emission<br> |
+| **Post-Conditions (Error)**    | 1. if `onReceivePacket` fails the application state is unchanged<br> 2. `packetReceipt is not written`<br> <br> 3. No Event Emission<br> | 1. `app.State(beforeReceivePacket)==app.State(afterReceivePacket)`<br> 2. `provableStore.get(packetReceiptPath(packet.channelDestId, packet.sequence))==null` <br> 3. Check No Event is Emitted<br> |
+               
+###### Pseudo-Code 
+
+The ICS-04 provides an example pseudo-code that enforce the above described conditions so that the following sequence of steps SHOULD occur for a packet to be received from module *1* on machine *A* to module *2* on machine *B*.
+
+>**Note:** We pass the address of the `relayer` that signed and submitted the packet to enable a module to optionally provide some rewards. This provides a foundation for fee payment, but can be used for other techniques as well (like calculating a leaderboard).
+
+```typescript
+function recvPacket(
+  packet: Packet,
+  proof: CommitmentProof,
+  proofHeight: Height,
+  relayer: string  
+  ) {
+
+    // Channel and Client Checks
+    channel = getChannel(packet.channelDestId)     
+    assert(channel !== null)
+    client = router.clients[channel.clientId]  
+    assert(client !== null)
+
+    // Check that counterparty channel id is as expected
+    assert(packet.sourceChannelId == channel.counterpartyChannelId) 
+    
+    // verify timeout
+    assert(packet.timeoutTimestamp === 0)  
+    assert(currentTimestamp() < packet.timeoutTimestamp)
+
+    // verify the packet receipt for this packet does not exist already 
+    packetReceipt = provableStore.get(packetReceiptPath(packet.channelDestId, packet.sequence))
+    abortTransactionUnless(packetReceipt === null)
+
+    //////// verify commitment 
+    
+    // 1. retrieve keys 
+    packetPath = packetCommitmentPath(packet.channelDestId, packet.sequence)
+    merklePath = applyPrefix(channel.keyPrefix, packetPath)
+    
+    // 2. reconstruct commit value based on the passed-in packet  
+    commit = commitV2Packet(packet) 
+    
+    // 3. call client verify memership 
+    assert(client.verifyMembership(
+        client.clientState 
+        proofHeight,
+        proof,
+        merklePath,
+        commit))
+
+    
+    // Executes Application logic ∀ Payload
+    payload=packet.data[0]
+    cbs = router.callbacks[payload.destPort]
+    ack,success = cbs.onReceivePacket(packet.channelDestId,payload,relayer,packet.sequence) // Note that payload includes the version. The application is required to inspect the version to route the data to the proper callback
```

Comment:
Discussing with @chatton about https://github.com/cosmos/ibc/issues/1141 we realised we do need the `packet.sequence` to be passed into the `onRecv` `onAck` `onTimeout` callbacks. 
Indeed, the ics20v2 the `packet.sequence` is used to store the packet in the `packetForwardPath` [e.g](https://github.com/cosmos/ibc/tree/main/spec/app/ics-020-fungible-token-transfer#:~:text=privateStore.set(packetForwardPath(forwarding.hops%5B0%5D.portId%2C%20forwarding.hops%5B0%5D.channelId%2C%20packetSequence)%2C%20packet))